guard/src/commands/validate/common.rs

```diff
@@ -0,0 +1,262 @@
+            },
+            message: each_failing_clause.msg.as_ref().map_or(
+                "".to_string(), |e| if !e.contains("DEFAULT") { e.clone() } else { "".to_string() })
+        })
+    }
+    else {
+        //
+        // This is crappy, but we are going to extract information from the retrieval error message
+        // see path_value.rs for retrieval error messages.
+        // TODO merge the query interface to retrieve partial results along with errored one ones and then
+        //      change this logic based on the reporting changes. Today we bail out for the first
+        //      retrieval error, fast fail semantics
+        //
+
+        //
+        // No from is how we indicate retrieval errors.
+        //
+        let (path, message) = each_failing_clause.msg.as_ref().map_or(("".to_string(), "".to_string()), |msg| {
+            match PATH_FROM_MSG.captures(msg) {
+                Some(cap) => (cap["path"].to_string(), msg.clone()),
+                None => ("".to_string(), msg.clone())
+            }
+        });
+
+        Ok(NameInfo {
+            rule: rule_name,
+            path,
+            provided: None,
+            expected: None,
+            comparison: None,
+            message
+        })
+    }
+}
+
+pub(super) fn colored_string(status: Option<Status>) -> ColoredString {
+    let status = match status {
+        Some(s) => s,
+        None => Status::SKIP,
+    };
+    match status {
+        Status::PASS => "PASS".green(),
+        Status::FAIL => "FAIL".red().bold(),
+        Status::SKIP => "SKIP".yellow().bold(),
+    }
+}
+
+pub(super) fn find_all_failing_clauses(context: &StatusContext) -> Vec<&StatusContext> {
+    let mut failed = Vec::with_capacity(context.children.len());
+    for each in &context.children {
+        if each.status.map_or(false, |s| s == Status::FAIL) {
+            match each.eval_type {
+                EvaluationType::Clause |
+                EvaluationType::BlockClause => {
+                    failed.push(each);
+                    if each.eval_type == EvaluationType::BlockClause {
+                        failed.extend(find_all_failing_clauses(each));
+                    }
+                },
+
+                EvaluationType::Filter |
+                EvaluationType::Condition => {
+                    continue;
+                },
+
+                _ => failed.extend(find_all_failing_clauses(each))
+            }
+        }
+    }
+    failed
+}
+
+pub(super) fn print_name_info<R, U, B>(
+    writer: &mut dyn Write,
+    info: &[NameInfo<'_>],
+    longest_rule_len: usize,
+    rules_file_name: &str,
+    data_file_name: &str,
+    retrieval_error: R,
+    unary_message: U,
+    binary_message: B) -> crate::rules::Result<()>
+    where R: Fn(&str, &str, &NameInfo<'_>) -> crate::rules::Result<String>,
+          U: Fn(&str, &str, &str, &NameInfo<'_>) -> crate::rules::Result<String>,
+          B: Fn(&str, &str, &str, &NameInfo<'_>) -> crate::rules::Result<String>
```

Comment:
unary and binary have the same api, any reason to declare them separately? Can we have a type "error" as `Fn(&str, &str, &NameInfo<'_> -> crate::rules::Result<String>)` and another type "message" as 
`Fn(&str, &str, &str, &NameInfo<'_>) -> crate::rules::Result<String>` instead of three types?

Reply:
https://users.rust-lang.org/t/rust-rejects-functions-of-the-same-type-when-a-fn-trait-bound-is-used/20746/2